pkg/agent/generate_secret_controller.go

```diff
@@ -0,0 +1,147 @@
+	hcAnnotation                  = "create-external-hub-kubeconfig"
+)
+
+type ExternalSecretController struct {
+	hubClient   client.Client
+	spokeClient client.Client
+	log         logr.Logger
+}
+
+// var ExternalSecretPredicateFunctions = predicate.Funcs{
+// 	CreateFunc: func(e event.CreateEvent) bool {
+// 		klog.Info("KLOG TRIGERRED BY CREATE")
+// 		return true
+// 	},
+// 	UpdateFunc: func(e event.UpdateEvent) bool {
+// 		klusterletOld := e.ObjectOld.(*operatorapiv1.Klusterlet)
+// 		klusterletNew := e.ObjectNew.(*operatorapiv1.Klusterlet)
+// 		if (klusterletOld.ObjectMeta.DeletionTimestamp.IsZero() != klusterletNew.ObjectMeta.DeletionTimestamp.IsZero()) {
+// 			klog.Info("TRIGERRED BY UPDATE KLOG")
+// 			klog.Info("Old has deletion time stamp?", klusterletOld.ObjectMeta.DeletionTimestamp.IsZero())
+// 			klog.Info("New has deletion time stamp?", klusterletNew.ObjectMeta.DeletionTimestamp.IsZero())
+// 		}
+// 		return klusterletOld.ObjectMeta.DeletionTimestamp.IsZero() != klusterletNew.ObjectMeta.DeletionTimestamp.IsZero()
+// 	},
+// 	DeleteFunc: func(e event.DeleteEvent) bool {
+// 		return false
+// 	},
+// }
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *ExternalSecretController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&operatorapiv1.Klusterlet{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
+		//WithEventFilter(ExternalSecretPredicateFunctions).
+		Complete(c)
+}
+
+// Reconcile updates the Hypershift addon status based on the Deployment status.
+func (c *ExternalSecretController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	c.log.Info(fmt.Sprintf("reconciling annotation for %s", req.Name))
+	defer c.log.Info(fmt.Sprintf("done reconciling annotation for %s", req.Name))
+
+	if !strings.Contains(req.Name, "klusterlet-") {
+		c.log.Info("klusterlet not from a managed cluster")
+		return ctrl.Result{}, nil //No need to error
+	}
+
+	//Get klusterlet
+	klusterletNamespaceNsn := types.NamespacedName{
+		Name:      req.Name,
+		Namespace: req.Namespace,
+	}
+	klusterlet := &operatorapiv1.Klusterlet{}
+	if err := c.spokeClient.Get(ctx, klusterletNamespaceNsn, klusterlet); err != nil {
+		c.log.Error(err, "Unable to get klusterlet")
+		return ctrl.Result{}, err
+	}
+
+	_, hostedClusterName, _ := strings.Cut(req.Name, "klusterlet-")
+
+	lo := &client.ListOptions{}
+	hostedClusters := &hyperv1beta1.HostedClusterList{}
+
+	// List the HostedCluster objects across all namespaces
+	if err := c.spokeClient.List(ctx, hostedClusters, lo); err != nil {
+		c.log.Error(err, "Unable to list hosted clusters in all namespaces")
+		return ctrl.Result{}, err
+	}
+
+	hostedClusterObj := &hyperv1beta1.HostedCluster{}
+	// Loop over the list of HostedCluster objects and find the one with the specified name
+	for index, hc := range hostedClusters.Items {
```

Comment:
Can we get the hosted cluster by name instead of listing all HCs and finding a match by name?

Reply:
@philipwu08 I thought of doing this, but I think hosted clusters can be in more than one namespace. I don't think the klusterlet provides the info for the namespace of the hosted cluster, that's why I look through all of them.